### PR TITLE
Match current taxonomy `dist` generation nuances

### PR DESCRIPTION
**Pull request by danielpgross**
This PR updates `dist` generation for the taxonomy to match nuances of the current behaviour.

* Values (JSON and TXT formats) are now sorted by name
* Attributes (JSON and TXT formats) are now sorted by name
* In `attributes.json`, values for a given attribute are now sorted
  * If `sorting: custom` is defined in the input file, the input sort order is preserved
    * A new `is_manually_sorted` field has been added to `Attribute`
  * Otherwise, `AlphanumericSorter` is used to sort them
    * `AlphanumericSorter` has been copied in from the current code
    * `ValueSorter` has been replaced with a single new class method on `Value` called `sort_by_localized_name`
* In `attributes.json`, extended attributes for a given attribute are now sorted by name
* In `categories.json`, attributes for a given category are now sorted by name
* In `categories.json`, children of a given category are now sorted by name

With these changes, I used a recursive folder diff tool to confirm that the new generated `dist` output exactly matches the old.

### CODE REVIEW

**Comment by danielpgross** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/Shopify/product-taxonomy/488?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#488** <a href="https://app.graphite.dev/github/pr/Shopify/product-taxonomy/488?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/Shopify/product-taxonomy/488?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#482** <a href="https://app.graphite.dev/github/pr/Shopify/product-taxonomy/482?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#480** <a href="https://app.graphite.dev/github/pr/Shopify/product-taxonomy/480?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: dev/lib/product_taxonomy/models/attribute.rb [hunk 1]
```diff
@@ -152,13 +180,13 @@ def to_json(locale: "en")
         "name" => name(locale:),
         "handle" => handle,
         "description" => description(locale:),
-        "extended_attributes" => extended_attributes.map do
+        "extended_attributes" => extended_attributes.sort_by(&:name).map do
```
Comment (tylerrowsell) [inline dev/lib/product_taxonomy/models/attribute.rb:183]: Should locale also be passed in here?

File: dev/lib/product_taxonomy/models/category.rb [hunk 2]
```diff
@@ -17,10 +17,13 @@ def load_from_source(source_data)
 
         # First pass: Create all nodes and add to index
         source_data.each do |item|
+          attributes = item["attributes"]
+            &.map { |attr_id| Attribute.find_by(friendly_id: attr_id) || attr_id }
+            &.sort_by { _1.is_a?(Attribute) ? _1.name : _1 }
```
Comment (tylerrowsell) [inline dev/lib/product_taxonomy/models/category.rb:22]: locale here too?

**Review by tylerrowsell — APPROVED** [review]
Changes look good. Just wondering if we need to consider locale more when sorting as well?

If so can we also add a test.

**Comment by danielpgross** [comment]
> Changes look good. Just wondering if we need to consider locale more when sorting as well?
> 
> If so can we also add a test. 

@tylerrowsell great question. In working on this, I actually checked each specific type of sorting in the old implementation and found that most of them _don't_ take the locale into account. Some examples:
https://github.com/Shopify/product-taxonomy/blob/4b326dcdddd27f4d0f76526af1b47f7794134985/app/models/attribute.rb#L4
https://github.com/Shopify/product-taxonomy/blob/4b326dcdddd27f4d0f76526af1b47f7794134985/app/models/value.rb#L4

You can see it very clearly if you look at attributes.txt for FR, where the localized values are clearly not sorted alphabetically (they're actually sorted alphabetically by the English name): https://github.com/Shopify/product-taxonomy/blob/33bdd359ef7a293ed24412d7a816d6386efb4a6f/dist/fr/attributes.txt

In fact, `values` for a given `Attribute`'s JSON was the only thing that did take locale into account.

I confirmed by diffing the localized `dist/` files between new and old -- the ordering is the same.

File: dev/lib/product_taxonomy/models/value.rb [hunk 3]
```diff
@@ -64,12 +64,31 @@ def reset
...
+
       private
 
       def longest_gid_length
         largest_id = hashed_by(:id).keys.max
         find_by(id: largest_id).gid.length
       end
+
+      def all_values_sorted(locale: "en")
+        all.sort_by { |value| [value.name(locale:) == "Other" ? 1 : 0, value.name(locale:), value.id] }
```
Comment (elsom25) [inline dev/lib/product_taxonomy/models/value.rb:90]: Should this force english and downcase like the other check?

**Review by elsom25 — APPROVED** [review]
Nothing blocking; a few minor nits/questions for consideration

File: dev/lib/product_taxonomy/models/extended_attribute.rb [hunk 4]
```diff
@@ -22,12 +22,13 @@ def initialize(name:, handle:, description:, friendly_id:, values_from:)
...
       values_from.add_extended_attribute(self) if values_from.is_a?(Attribute)
       super(
-        id: nil,
+        id: values_from.is_a?(Attribute) ? values_from.id : nil,
         name:,
         handle:,
         description:,
         friendly_id:,
         values: values_from.is_a?(Attribute) ? values_from.values : nil,
+        is_manually_sorted: values_from.is_a?(Attribute) ? values_from.manually_sorted? : nil,
```
[the remark this answers is not in the document]
Comment (elsom25) [reply dev/lib/product_taxonomy/models/extended_attribute.rb:31]:
```suggestion
        is_manually_sorted: values_from.try(:manually_sorted?) || false,
```

File: dev/lib/product_taxonomy/models/value.rb (hunk 3 above)
Comment (elsom25) [reply dev/lib/product_taxonomy/models/value.rb:90]:
```suggestion
        all.sort_by { |value| [value.name(locale: "en").downcase == "other" ? 1 : 0, value.name(locale:), value.id] }
```

**Pull request merged**